lib/fog/ovirt/compute/v4.rb

```diff
@@ -0,0 +1,161 @@
+              if key == :provisioned_size
+                opts[:size] = value
+                next
+              end
+
+              # ignore nil values
+              next if value.nil?
+
+              if key != :data_center && value.respond_to?(:href) && value.href && value.respond_to?(:id)
+                opts[key] = value.id
+              end
+
+              opts[key] ||= get_attr_value(value, opts)
+            end
+            opts
+          end
+          # rubocop:enable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+          # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity
+          def get_attr_value(value, opts)
+            case value
+            when OvirtSDK4::TemplateVersion, Array, Hash, OvirtSDK4::List, OvirtSDK4::DataCenter
+              value
+            when OvirtSDK4::Mac
+              value.address
+            when OvirtSDK4::Cpu
+              opts[:cores] = value.topology.nil? ? nil : value.topology.cores
+            when OvirtSDK4::Display
+              subject = value.certificate.subject if value.certificate
+              {
+                :type => value.type,
+                :address => value.address,
+                :port => value.port,
+                :secure_port => value.secure_port,
+                :subject => subject,
+                :monitors => value.monitors
+              }
+            else
+              value.to_s.strip
+            end
+          end
+          # rubocop:enable Metrics/AbcSize, Metrics/CyclomaticComplexity
+        end
+
+        class Mock
+          include Shared
+
+          def initialize(_options = {})
+            require "ovirtsdk4"
+          end
+
+          private
+
+          def client
+            return @client if defined?(@client)
+          end
+
+          # read mocks xml
+          def read_xml(file_name)
+            file_path = File.join(__dir__, "../requests", "compute/v4", "mock_files", file_name)
+            File.read(file_path)
+          end
+        end
+
+        class Real
+          include Shared
+
+          # rubocop:disable Metrics/AbcSize
+          def initialize(options = {})
+            require "ovirtsdk4"
+            username   = options[:ovirt_username]
+            password   = options[:ovirt_password]
+            server     = options[:ovirt_server]
+            port       = options[:ovirt_port]       || 8080
+            api_path   = options[:ovirt_api_path]   || "/api"
+            url        = options[:ovirt_url]        || "#{@scheme}://#{server}:#{port}#{api_path}"
+
+            connection_opts = {
+              :url      => url,
+              :username => username,
+              :password => password
+            }
+
+            @datacenter = options[:ovirt_datacenter]
+            connection_opts[:ca_file]  = options[:ca_file]
+            connection_opts[:ca_certs] = [OpenSSL::X509::Certificate.new(options.delete(:public_key))] if options[:public_key].present?
```

Comment:
I guess we should not modify the `options` hash and use `options[:public_key]` instead of `options.delete(:public_key)`